add helper for dealing with video dimensions

Jalle19 opened the pull request:
This is a first attempt at hopefully consolidating the logic behind determining whether a video is SD or HD etc. It replaces `CStreamDetails::VideoDimsToResolutionDescription` with a new helper class `CVideoDimensions` which provides the same functionality and more.

This also changes the logic for converting width and height to e.g. "720" to use a list of boundaries that each "descriptor" can have, e.g. for a file to be "720" its width must be between 960 and 1280 while its height must be between 540, 720. The way this is implemented means it should be easy to create unit tests to make sure this crap doesn't keep breaking everyone someone so much as looks at the code.

The use of boundaries to specify what "descriptor" a video item has can also be used backwards, e.g. in order to find all videos having a resolution matching "720". Before that was not possible because the boundaries weren't defined anywhere. This is needed in order to implement some of the suggestions in https://github.com/xbmc/xbmc/pull/6611

RFC:
- should we try to bake this into e.g. `CStreamDetailsVideo` instead? Somewhere else? There are lots of classes that have just two members (widht and height) so there's probably more room for consolidation
- The naming scheme could probably be more descriptive (no pun intended)

TODO:
- Not sure where `GetFidelity` could be used at the moment, I need to check the code for the lists (where the SD/HD icon is shown).

@mkortstiege @Montellese @topfs2 @fritsch

Jalle19 committed:
introduce a helper for categorizing video dimensions into meaningful
descriptors like "720" and "1080". This also changes the logic for
determining what descriptor a video has from just looking at the width
and height to checking that they are between the defined boundaries

MartijnKaijser commented:
be careful with 3D-SBS videos. They can be very tricky to handle and you haven't taken care of those yet.

Jalle19 commented:
@MartijnKaijser I adjusted the boundaries
@Montellese any guide to how I can run the gtests in VS?

Montellese commented:
I'll review later but I like the approach.

Concerning unit tests you need to switch the configuration in VS to "Debug Testsuite", build that and then run the "Kodi" project. To be able to see the output I usually put a breakpoint in `tests/xbmc-test.cpp`.

ace20022 commented:
Currently there doesn't seem to be a need for an (non static) object. Are you planning to store it in the videoinfotag?

Jalle19 commented:
@ace2022 I had something like that in mind, yes. Gotta take a closer look and decide where it would best fit in.

Montellese commented on xbmc/video/VideoDimensions.cpp:
```diff
@@ -0,0 +1,96 @@
+{
+  std::string descriptor = GetDescriptor();
+
+  for (const auto &fidelity : fidelities)
```
Why not use `fidelities.find(GetDescriptor())`?

Jalle19 replied:
I got some random error during compilation but I guess it must have been something else. I'll change it.

Montellese commented on xbmc/video/VideoDimensions.cpp:
```diff
@@ -0,0 +1,96 @@
+
+Boundary CVideoDimensions::GetBoundary(const std::string &descriptor)
+{
+  return CVideoDimensions::boundaries[descriptor];
```
No check if the `descriptor` exists in the map?
And you should be able to drop the `CVideoDimensions::`.

Jalle19 replied:
I omitted it on purpose for now, I'm not sure what to return if no boundary is found. I'd like to avoid dealing with pointers if possible.

Montellese replied:
Maybe define a static empty boundary where all values are 0 and return that if no boundary was found? Then the calling code can check against that empty boundary.

Montellese commented on xbmc/video/VideoDimensions.cpp:
```diff
@@ -0,0 +1,96 @@
+{
+  std::vector<std::string> descriptors;
+
+  for (const auto &boundary : CVideoDimensions::boundaries)
```
You should be able to drop the `CVideoDimensions::`.

Montellese commented on xbmc/video/VideoDimensions.h:
```diff
@@ -0,0 +1,73 @@
+#include <vector>
+#include <utility>
+
+typedef std::pair<std::pair<int, int>, std::pair<int, int>> Boundary;
```
I'd prefer a struct with more obvious names over a combination of `std::pair` inside `std::pair`. You'd have a struct with `min` and `max` members and one with `width` and `height` members of the previous struct.

Montellese commented on xbmc/video/VideoDimensions.h:
```diff
@@ -0,0 +1,73 @@
+  /*
+   * Maps a descriptor (e.g. "720") to a fidelity (e.g. "HD")
+   */
+  static std::map<std::string, std::string> fidelities;
```
Is there any reason you need two maps? They both use the exact same keys so if you'd use a struct you could store all the information in one structure and use one map.

Jalle19 replied:
Good idea regarding the struct, it's easier to document too.

Jalle19 replied:
In a second thought, the "fidelity" is not related to a "boundary" at all, so without changing the name of `Boundary` I wouldn't want to get rid of that second map.

Jalle19 replied:
On a third thought, I came up with something else. Will push some changes soon.

Montellese replied:
You can have a struct `BoundaryLimits` with `min` and `max`, a struct `Boundary` with two `BoundaryLimits` members called `width` and `height` and another struct `Dimensions` with a `Boundary boundary` member and a `std::string fidelity` member. Then you have a map `std::map<std::string, Dimensions> dimensions` or something like that. Obviously all these names are adjustable.

Montellese commented:
Would it make sense to configure these boundary values somewhere in an XML file so that we don't have to change the code to change the boundaries?

mkortstiege commented:
@Jalle19 pick xcode sync from https://github.com/mkortstiege/xbmc/commit/185afa8d1a06eb23ead95d18e035c43512eb192e. Please do not remove the unrelated changes. This will bring the xcode project file in sync once for all (hopefully).

Jalle19 commented:
@Montellese I refactored it so that each resolution can be represented by a struct (`SResolution`). A resolution has a descriptor, a fidelity and a `SResolutionBoundary` struct, which represents the dimensional limits of the resolution. The corresponding `SResolution` is stored as a member during construction, and the various methods just return parts of the struct (so that the caller doesn't need to bother with the struct's structure).

Do you want to drop the static `GetResolutionData` and just make the map public? The benefit of the current approach is that the map is effectively read-only.

Montellese commented:
Yup looks much better IMO and easier to read/understand. I would keep the static `GetResolutionData` because we don't want anyone to mess around with the static list of resolutions.

da-anda commented:
what about FullTAB/OU and some awkward 8:3 or 3:8 FullOU/SBS material (some older 3D stuff with two 4:3 frames next to each other)? Wouldn't it make sense to pass some flag to the helper function if the video is 3D or not? I mean when we have the dimensions from the metadata we should as well have the info if it's 3D or not. I tried to streamline 3D detection and store the result in streamdetails in this branch (https://github.com/da-anda/xbmc/commits/3d-centralize-detection), but it went a bit over my head, so somebody with more insight would have to take it over. I'm happy to assist though. In my branch I also try to detect if a movie is 3D or not based on it's resolution (which worked fine for me in the spot I added it, but  elupus wanted the logic in a different place where I couldn't solve it) - so maybe this logic could also be integrated to the new helper from this PR?

Montellese commented:
IMO let's first get this done as a replacement for the logic we already have in `CStreamDetails` otherwise it won't be finished (and maybe even abandoned) for a while.

da-anda commented:
sure, didn't mean to block this one - only tried to mention that to handle 3D stuff correctly some extra work might be needed

mkortstiege commented:
`GetFidelity` could be used within smart playlists as a really simple filter.

Jalle19 commented:
@da-anda @Montellese I was thinking we could combine this new `CVideoDimensions` class into `CStreamDetailVideo`, that's the class where all code gets the width and height from, and it contains the stereoscopic mode too.

@mkortstiege indeed, we just need a method for grabbing the `Boundary` based on a fidelity.

da-anda commented:
@Jalle19 well, streamdetails atm don't necessarily contain a stereo_mode value even though the video is 3D. That's what I tried to fix in my branch (make streamdetails the only source for the stereo_mode value). But yeah, once this is fixed it might work.

Jalle19 commented:
Okay, did you figure out why it's not always set?

Jalle19 commented:
@mkortstiege do you need anything other than `GetResolutionData` to implement a per descriptor filter?

To all: should we rename "descriptor"? Maybe "quality" would make more sense.

Jalle19 commented:
@Montellese regarding defining boundaries in XML, I don't think that improves anything since someone still has to change a file. I doubt it makes any sense to make the boundaries user configurable, they are supposed to be mathematical facts (e.g. 360x240 cannot ever be called "HD").

Montellese commented:
@Jalle19: Yeah I just don't like these fixed lists in code but it doesn't really matter to me.

Concerning smartplaylists we just need access to the boundary values (to be able to generate a proper SQL query) and the name (to be able to let the user choose from a list).

mkortstiege commented:
+1 for `quality` instead of `descriptor`. And what @Montellese wrote.

Jalle19 commented:
Okay, in that case I think it's best if we just let people rely on `GetResolutionData`. The problem with having two methods, e.g. `GetDescriptors()` (which returns "480", "720" etc.) and a separate static one called `GetBoundary(string descriptor)` is that the latter would have to do error checking in case someone passes in a non-existing descriptor. Or what do you think? Can we just skip the error checking?

Jalle19 commented:
To keep the ball moving I renamed some things (see commit message) and introduced methods for getting the boundaries based on quality (e.g. "720") or fidelity (e.g. "HD"). These could be used for filtering in #6611 (if reopened).

Jalle19 commented:
@mkortstiege squashed and rebased, but unfortunately I fubar'd the xcode sync so you'll have to redo that one.

Jalle19 commented:
jenkins build this please

Jalle19 commented:
jenkins build this please

Jalle19 commented:
@mkortstiege your call if we should merge this. At least it seems to build now.

mkortstiege commented:
I'd say merge but would like to have another +1.

AlwinEsch commented on xbmc/video/VideoDimensions.cpp:
```diff
@@ -0,0 +1,118 @@
+
+CVideoDimensions::CVideoDimensions(int width, int height) :
+m_width(width),
+m_height(height)
```
Here are spaces missing.

mkortstiege commented:
@Jalle19 just checked again. You're sure this can be used to filter without relying on predefined values?

Jalle19 commented:
@mkortstiege what do you mean?

Jalle19 committed:
[tests] added a massive test suite based on multiple complete video
libraries (plus some users reports on the forum) and adjusted the
boundaries to make the tests pass

Jalle19 commented:
Okay so I've added tests based on my and @kib's libraries, if anyone has some 3D and/or 4K material please speak up so we can integrate those too.

Jalle19 committed:
[smartplaylist] changed the video resolution filter to a browseable
filter where you select the resolution(s) you want instead of having
to enter a specific video height

Jalle19 commented:
@mkortstiege I added the smartplaylist part now, can you take a look? I also have code for filtering by "fidelity", i.e. just "SD" or "HD" but I thought I'd save that for a separate PR.

This should fix all the issues reported in http://forum.kodi.tv/showthread.php?tid=109809

Jalle19 commented:
jenkins build this please

da-anda commented on xbmc/video/VideoDimensions.cpp:
```diff
@@ -0,0 +1,118 @@
+  { "576", "SD", { { 0, 768 }, { 480, 600 } } },
+  // 720p can be anything from 960x720 (1.33:1) to 1280x464 (2.70:1, e.g. Ben Hur)
+  { "720", "HD", { { 768, 1280 }, { 460, 720 } } },
+  // Full-SBS 3D can be 3840x1080, some weird shit can be 1080x1040
```
don't forget about Full-TAB which would be 1920x2160 (also in your test cases). Also there might be SBS or OU 4:3 SD material, like it can be found here: http://www.3dtv.at/Movies/Index_en.aspx (the skydive f.e., which currently has scaling issues IIRC, but that's a different story)

Jalle19 committed:
[squash] fix invalid #include

Jalle19 commented:
jenkins build this please

Jalle19 commented:
@mkortstiege regarding your IRC comment about the playlist filter I'm not sure how to fix it since I'm unfamiliar with the code. I would like to limit the rule so you can't choose an operator other than "is" and you can't select multiple resolutions.

Montellese commented:
@Jalle19: That would probably be possible BUT you can't stop users from defining multiple values or using another operator in the XML files. But what is the problem with defining multiple values? What if I want all movies that are 720p or 1080p?

MartijnKaijser commented:
@Jalle19 it would be great if this could go in at some point. Is there any info you need?

Jalle19 commented:
@MartijnKaijser it will need some more refactoring before it's ready.

Rechi commented:
@Jalle19 this needs a rebase

MartijnKaijser commented:
I will close this PR. If there's interest please open a new one (with required changes if needed).
